iroh-relay/src/server/streams.rs

```diff
@@ -190,4 +217,307 @@ impl AsyncWrite for MaybeTlsStream {
             MaybeTlsStream::Test(ref mut s) => Pin::new(s).poll_write_vectored(cx, bufs),
         }
     }
+
+    fn is_write_vectored(&self) -> bool {
+        match self {
+            MaybeTlsStream::Plain(s) => s.is_write_vectored(),
+            MaybeTlsStream::Tls(s) => s.is_write_vectored(),
+            #[cfg(test)]
+            MaybeTlsStream::Test(s) => s.is_write_vectored(),
+        }
+    }
+}
+
+/// Rate limiter for reading from a [`RelayedStream`].
+///
+/// The writes to the sink are not rate limited.
+///
+/// This potentially buffers one frame if the rate limiter does not allows this frame.
+/// While the frame is buffered the undernlying stream is no longer polled.
+#[derive(Debug)]
+pub(crate) struct RateLimited<S> {
+    inner: S,
+    bucket: Option<Bucket>,
+    bucket_refilled: Option<Pin<Box<time::Sleep>>>,
+    /// Keeps track if this stream was ever rate-limited.
+    limited_once: bool,
+    metrics: Arc<Metrics>,
+}
+
+#[derive(Debug)]
+struct Bucket {
+    // The current bucket fill
+    fill: i64,
+    // The maximum bucket fill
+    max: i64,
+    // The bucket's last fill time
+    last_fill: time::Instant,
+    // Interval length of one refill
+    refill_period: time::Duration,
+    // How much we re-fill per refill period
+    refill: i64,
+}
+
+#[allow(missing_docs)]
+#[derive(Debug, Snafu)]
+pub struct InvalidBucketConfig {
+    backtrace: Option<Backtrace>,
+    #[snafu(implicit)]
+    span_trace: n0_snafu::SpanTrace,
+    max: i64,
+    bytes_per_second: i64,
+    refill_period: time::Duration,
+}
+
+impl Bucket {
+    fn new(
+        max: i64,
+        bytes_per_second: i64,
+        refill_period: time::Duration,
+    ) -> Result<Self, InvalidBucketConfig> {
+        // milliseconds is the tokio timer resolution
+        snafu::ensure!(
+            max > 0 && bytes_per_second > 0 && refill_period.as_millis() != 0,
+            InvalidBucketConfigSnafu {
+                max,
+                bytes_per_second,
+                refill_period,
+            },
+        );
+        Ok(Self {
+            fill: max,
+            max,
+            last_fill: time::Instant::now(),
+            refill_period,
+            refill: bytes_per_second * refill_period.as_millis() as i64 / 1000,
+        })
+    }
+
+    fn update_state(&mut self) {
+        let now = time::Instant::now();
+        let refill_periods = now.saturating_duration_since(self.last_fill).as_millis() as u32
+            / self.refill_period.as_millis() as u32;
+        if refill_periods == 0 {
+            // Nothing to do - we won't refill yet
+            return;
+        }
+
+        self.fill += refill_periods as i64 * self.refill;
+        self.fill = std::cmp::min(self.fill, self.max);
+        self.last_fill += self.refill_period * refill_periods;
+    }
+
+    fn consume(&mut self, bytes: i64) -> Result<(), time::Instant> {
+        self.update_state();
+
+        self.fill -= bytes;
```

Comment:
can this not underflow/wrap-around?

Reply:
Thank you for these comments, I've added some tests & made the integer arithmetic more robust w.r.t. close-to-bounds values for configurable values (`bytes_per_second` being super high or the value passed to `consume` being insanely high).

Also made sure to check that we never div by zero etc.
https://github.com/n0-computer/iroh/pull/3384/commits/46e8cd6a7bd1707ad432ae61be01e750222a89ee